designs/component-config.md

```diff
@@ -0,0 +1,312 @@
+
+```golang
+// pkg/api/config/v1alpha1/types.go
+package v1alpha1
+
+import (
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	metav1alpha1 "sigs.k8s.io/controller-runtime/pkg/apis/meta/v1alpha1"
+)
+
+// DefaultControllerConfiguration is the Schema for the DefaultControllerConfigurations API
+type DefaultControllerConfiguration struct {
+	metav1.TypeMeta `json:",inline"`
+
+	Spec   metav1alpha1.RuntimeConfig   `json:"spec,omitempty"`
+}
+```
+
+This would allow a controller author to use this struct with any config that supports the json/yaml structure. For example a controller author could define their `Kind` as `FoobarControllerConfiguration` and have it defined as the following.
+
+```yaml
+# config.yaml
+apiVersion: somedomain.io/v1alpha1
+kind: FoobarControllerConfiguration
+spec:
+  port: 9443
+  metricsBindAddress: ":8080"
+  leaderElection:
+    leaderElect: false
+```
+
+Given the following config and `DefaultControllerConfiguration` we'd be able to initialize the controller using the following.
+
+
+```golang
+mgr, err := ctrl.NewManagerFromComponentConfig(ctrl.GetConfigOrDie(), scheme, configname, &defaultv1alpha1.DefaultControllerConfiguration{})
+if err != nil {
+	// ...
+}
+```
+
+The above example uses `configname` which is the name of the file to load the configuration from and uses `scheme` to get the specific serializer, eg `serializer.NewCodecFactory(scheme)`. This will allow the configuration to be unmarshalled into the `runtime.Object` type and passed into the 
+`ctrl.NewManagerFromComponentConfig()` as a `ManagerConfiguration` interface.
+
+#### Using Flags w/ ComponentConfig
+
+Since this design still requires setting up the initial `ComponentConfig` type and passing in a pointer to `ctrl.NewFromComponentConfig()` if you want to allow for the use of flags your controller can use any of the different flagging interfaces. eg [`flag`](https://golang.org/pkg/flag/), [`pflag`](https://godoc.org/github.com/spf13/pflag), [`flagnum`](https://godoc.org/github.com/luci/luci-go/common/flag/flagenum) and set values on the `ComponentConfig` type prior to passing the pointer into the `ctrl.NewFromComponentConfig()`.
+
+```golang
+leaderElect := true
+
+config := &defaultv1alpha1.DefaultControllerConfiguration{
+	Spec: metav1alpha1.RuntimeConfig{
+		LeaderElection: configv1alpha1.LeaderElectionConfiguration{
+			LeaderElect: &leaderElect,
+		},
+	},
+}
+mgr, err := ctrl.NewManagerFromComponentConfig(ctrl.GetConfigOrDie(), scheme, configname, )
+if err != nil {
+	// ...
+}
+```
+
+#### Kubebuilder Scaffolding Example
+
+Within expanded in a separate design _(link once created)_ this will require controller authors to generate a type that implements the `ManagerConfiguration` interface. The following is a sample of what this looks like:
```

Comment:
Just a note, that within this design we would need to expose the component config from the manager then so that a user could retrieve the custom type to make the custom choices they would like to make I think?

Reply:
@shawn-hurley given that you pass in a pointer the `decodeComponentConfigFileInto()` can hydrate the ComponentConfig type. In my examples this is implemented as:

```
// decodeComponentConfigFileInto will load the component config into the runtime.Object
func decodeComponentConfigFileInto(codecs serializer.CodecFactory, filePath string, obj runtime.Object) error {
	content, err := ioutil.ReadFile(filePath)
	if err != nil {
		return err
	}
	// Regardless of if the bytes are of any external version,
	// it will be read successfully and converted into the internal version
	return runtime.DecodeInto(codecs.UniversalDecoder(), content, obj)
}
```

Does that cover what you are saying? I could see someone then passing that object into the controllers for use of the configs from the reconcilers (at least in kubebuilder's case). Potentially it makes sense to also make `GetComponentConfig()` on the `Manager` interface so that everything is self contained.

Thoughts?

Reply:
I was thinking about how a user could use their custom component config in their reconciler, I was thinking of someting like  `GetComonentConfig(<Interface>)` such that they could pass in their component config and we would populate it? 

IDK WDYT?

Reply:
Ah, since the `NewFromComponentConfig()` func does this when you call it with a custom type or the default type you'll already have that populated within your `main.go` or where ever you init'ed the manager, so passing it into the reconciler as another struct tag seems like it would work the same? If it was just `GetComponentConfig() ManagerConfiguration` I could understand but I don't think providing another want to unmarshal is necessary.

Reply:
The example looks something like this:

```
componentConfig := &custom.HighlyCustomizedControllerConfiguration{}
mgr, err := ctrl.NewManagerFromComponentConfig(ctrl.GetConfigOrDie(), scheme, configname, componentConfig)

fmt.Printf("%+v", componentConfig)
fmt.Printf("GetPort() %d", componentConfig.GetPort())
```

and this will return the information from the config passed in via `configname`.